epochStart/metachain/legacySystemSCs.go

```diff
@@ -0,0 +1,1319 @@
+	for shId, validatorsInfoSlice := range validatorsInfoMap {
+		newList := make([]*state.ValidatorInfo, 0, len(validatorsInfoSlice))
+		deleteCalled := false
+
+		for _, validatorInfo := range validatorsInfoSlice {
+			if vInfo.WasEligibleInCurrentEpoch(validatorInfo) {
+				newList = append(newList, validatorInfo)
+				continue
+			}
+
+			err := s.stakingDataProvider.FillValidatorInfo(validatorInfo.PublicKey)
+			if err != nil {
+				deleteCalled = true
+
+				log.Error("fillStakingDataForNonEligible", "error", err)
+				if len(validatorInfo.List) > 0 {
+					return err
+				}
+
+				err = s.peerAccountsDB.RemoveAccount(validatorInfo.PublicKey)
+				if err != nil {
+					log.Error("fillStakingDataForNonEligible removeAccount", "error", err)
+				}
+
+				continue
+			}
+
+			newList = append(newList, validatorInfo)
+		}
+
+		if deleteCalled {
+			validatorsInfoMap[shId] = newList
+		}
+	}
+
+	return nil
+}
+
+func (s *legacySystemSCProcessor) prepareStakingDataForEligibleNodes(validatorsInfoMap map[uint32][]*state.ValidatorInfo) error {
+	eligibleNodes := s.getEligibleNodeKeys(validatorsInfoMap)
+	return s.prepareStakingData(eligibleNodes)
+}
+
+func (s *legacySystemSCProcessor) unStakeNonEligibleNodesWithNotEnoughFunds(validatorsInfoMap map[uint32][]*state.ValidatorInfo, epoch uint32) (uint32, error) {
+	err := s.fillStakingDataForNonEligible(validatorsInfoMap)
+	if err != nil {
+		return 0, err
+	}
+
+	return s.unStakeNodesWithNotEnoughFunds(validatorsInfoMap, epoch)
+}
+
+func (s *legacySystemSCProcessor) prepareStakingData(nodeKeys map[uint32][][]byte) error {
+	sw := core.NewStopWatch()
+	sw.Start("prepareStakingDataForRewards")
+	defer func() {
+		sw.Stop("prepareStakingDataForRewards")
+		log.Debug("systemSCProcessor.prepareStakingDataForRewards time measurements", sw.GetMeasurements()...)
+	}()
+
+	return s.stakingDataProvider.PrepareStakingData(nodeKeys)
+}
+
+func (s *legacySystemSCProcessor) getEligibleNodeKeys(
+	validatorsInfoMap map[uint32][]*state.ValidatorInfo,
+) map[uint32][][]byte {
+	eligibleNodesKeys := make(map[uint32][][]byte)
+	for shardID, validatorsInfoSlice := range validatorsInfoMap {
+		eligibleNodesKeys[shardID] = make([][]byte, 0, s.nodesConfigProvider.ConsensusGroupSize(shardID))
+		for _, validatorInfo := range validatorsInfoSlice {
+			if vInfo.WasEligibleInCurrentEpoch(validatorInfo) {
+				eligibleNodesKeys[shardID] = append(eligibleNodesKeys[shardID], validatorInfo.PublicKey)
+			}
+		}
+	}
+
+	return eligibleNodesKeys
+}
+
+// ProcessDelegationRewards will process the rewards which are directed towards the delegation system smart contracts
+func (s *legacySystemSCProcessor) ProcessDelegationRewards(
+	miniBlocks block.MiniBlockSlice,
+	txCache epochStart.TransactionCacher,
+) error {
+	if txCache == nil {
+		return epochStart.ErrNilLocalTxCache
+	}
+
+	rwdMb := getRewardsMiniBlockForMeta(miniBlocks)
```

Comment:
`getRewardsMiniBlockForMeta` is only called in legacy. Perhaps move the func in this file?

Reply:
moved